source/main.cpp

```diff
@@ -365,160 +357,34 @@ WMMSettingsView::~WMMSettingsView()
 	delete fAuthorView;
 
 	delete Icon;
-
-	SaveSettings( &Settings);
 }
 
 //---------------------------------------------------
-//	Attached to window - set siblings target to this
+//	Attached to window - set siblings target
 //---------------------------------------------------
 void
 WMMSettingsView::AttachedToWindow()
 {
-	slidRadius->SetTarget( this);
-	slidCircles->SetTarget( this);
-	slidLineWidth->SetTarget( this);
-	slidLineSpace->SetTarget( this);
-	slidPulseRate->SetTarget( this);
-	Window()->SetPulseRate( Settings.pulse);
+	slidRadius->SetTarget(fDemoView);
+	slidCircles->SetTarget(fDemoView);
+	slidLineWidth->SetTarget(fDemoView);
+	slidLineSpace->SetTarget(fDemoView);
+	slidPulseRate->SetTarget(fDemoView);
 }
 
 //---------------------------------------------------
 //	Draw view
 //---------------------------------------------------
 void
-WMMSettingsView::Draw( BRect updateRect)
-{
-	demoRect = fDemoView->Bounds();
-
-	//demo background
-	fDemoView->SetHighColor(255, 255, 255, 255);
-	fDemoView->FillRoundRect(demoRect, 2, 2);
-	fDemoView->SetHighColor(0, 0, 0, 255);
-	fDemoView->StrokeRoundRect(demoRect, 2, 2);
-
-	InitDemo();
-}
-
-//---------------------------------------------------
-//	Draw demo
-//---------------------------------------------------
-void
-WMMSettingsView::DrawDemo()
+WMMSettingsView::Draw(BRect updateRect)
 {
-	float right = demoRect.right, bottom = demoRect.bottom;
-
-	if( demoCircle < Settings.circles)
-	{
-		float oldPen = fDemoView->PenSize();
-		fDemoView->SetPenSize(Settings.lwidth);
-		BRegion region = BRegion(BRect(1, 1, right - 1, bottom - 1));
-		fDemoView->ConstrainClippingRegion(&region);
-
-		fDemoView->SetHighColor(demoColor);
-		fDemoView->StrokeEllipse(BPoint(right / 2, bottom / 2), demoRadius, demoRadius);
-		demoRadius += ( Settings.lwidth / 2) + Settings.lspace;
-		demoColor.alpha += demoAlphaMod;
-		demoCircle++;
-
-		fDemoView->ConstrainClippingRegion(NULL);
-		fDemoView->SetPenSize(oldPen);
-	}
-	else
-	{
-		//demo background
-		fDemoView->SetHighColor(255, 255, 255, 255);
-		fDemoView->FillRoundRect(BRect(1, 1, right - 1, bottom - 1), 2, 2);
-		//SetHighColor( 0, 0, 0, 255);
-		//StrokeRoundRect( demoRect, 2, 2);
-
-		InitDemo();
-	}
+	fDemoView->Invalidate();
```

Comment:
Here is the code (based on the original code) that initially draws the empty white rectangular box in Draw() and draws the circles one per pulse in DrawDemo(), which also erases the area within the box when the preset number of circles are drawn. How would you change it so that all the drawing occurs in Draw()? I guess it would require ClipToPicture() to tell Draw() to only redraw the next circle? And make several state variables local in DrawDemo() to class variables so that Draw() can use them to determine which way to draw (drawing the empty box, drawing a circle, or erasing the area within the box)?

void
DemoView::Draw(BRect updateRect)
{
	//demo background
	SetHighColor(255, 255, 255, 255);
	FillRoundRect(updateRect, 2, 2);
	SetHighColor(0, 0, 0, 255);
	StrokeRoundRect(updateRect, 2, 2);

	InitDemo();
}

void
DemoView::DrawDemo()
{
	BRect rect = Bounds();
	float x = rect.right / 2;
	float y = rect.bottom / 2;

	rect.left++;
	rect.top++;
	rect.right--;
	rect.bottom--;

	if (demoCircle < Settings.circles)
	{
		float oldPen = PenSize();
		SetPenSize(Settings.lwidth);
		BRegion region = BRegion(rect);
		ConstrainClippingRegion(&region);

		SetHighColor(demoColor);
		StrokeEllipse(BPoint(x, y), demoRadius, demoRadius);
		demoRadius += Settings.lwidth / 2 + Settings.lspace;
		demoColor.alpha += demoAlphaMod;
		demoCircle++;

		ConstrainClippingRegion(NULL);
		SetPenSize(oldPen);
	}
	else
	{
		//demo background
		SetHighColor(255, 255, 255, 255);
		FillRoundRect(rect, 2, 2);
		//SetHighColor( 0, 0, 0, 255);
		//StrokeRoundRect( demoRect, 2, 2);

		InitDemo();
	}
}